File: 总题库/301.RemoveInvalidParentheses.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        def isValid(s: str) -> bool:
            cnt = 0
            for character in s:
                if character == '(':
                    cnt += 1
                elif character == ')':
                    if cnt < 1: return False
                    else:
                        cnt -= 1
            return cnt == 0
        
        level = {s}
        while level:
            valid = list(filter(isValid, level))
            if valid: return valid
            nextLevel = set()

            for item in level:
                for i in range(len(item)):
                    if item[i] in '()':
                        nextLevel.add(item[:i] + item[i + 1:])
            level = nextLevel

File: 总题库/301.RemoveInvalidParentheses.cpp
// BFS
class Solution {
public:
    bool isValid(string s) {
        int cnt = 0;
        for (char c : s) {
            if (c == '(') ++cnt;
            else if (c == ')') {
                if (cnt < 1) return false;
                --cnt;
            }
        }
        return cnt == 0;
    }

    vector<string> removeInvalidParentheses(string s) {
        set<string> level;
        vector<string> res;
        level.insert(s);
        while (level.size()) {
            for (auto iter = level.begin(); iter != level.end(); ++iter) {
                if (isValid(*iter)) {
                    res.push_back(*iter);
                }
            }
            if (res.size() != 0) break;
            set<string>nextLevel;
            for (auto iter = level.begin(); iter != level.end(); ++iter) {
                string temp = *iter;
                for (int i = 0; i < temp.size(); ++i) {
                    if (temp[i] == '(' || temp[i] == ')') {
                        nextLevel.insert(temp.substr(0, i) + temp.substr(i + 1));
                    }
                }
            }
            level = nextLevel;
        }
        return res;
    }
};

// DFS